### PR TITLE
ARROW-3762: [C++/Python] Support reading Parquet BYTE_ARRAY columns containing over 2GB of data

### PR DESCRIPTION
**Pull request by wesm**
This patch ended up being a bit more of a bloodbath than I planned: please accept my apologies.

Associated changes in this patch:

* Split up builder.h/builder.cc into a new arrow/array directory. Public arrow/builder.h API preserved. I think this code is going to keep growing more specialized components, so I think we should get out ahead of it by having a subdirectory to contain files related to implementation details
* Implement ChunkedBinaryBuilder, ChunkedStringBuilder classes, add tests and benchmarks
* Deprecate parquet::arrow methods returning Array
* Allow implicit construction of Datum from its variant types (makes for a lot nicer syntax)

As far as what code to review, focus efforts on

* src/parquet/arrow
* src/arrow/array/builder_binary.h/cc, array-binary-test.cc, builder-benchmark
* src/arrow/compute changes
* Python changes

I'm going to tackle ARROW-2970 which should not be complicated after this patch; I will submit that as a PR after this is reviews and merged.

### CODE REVIEW

**Comment by wesm** [comment]
cc @xhochy @pitrou

File: cpp/src/arrow/CMakeLists.txt [hunk 1]
```diff
@@ -69,6 +66,14 @@ set(ARROW_SRCS
   type.cc
   visitor.cc
 
+  columnar/builder_adaptive.cc
```
Comment (pitrou) [inline cpp/src/arrow/CMakeLists.txt]: Hmm... Honestly "columnar" doesn't mean anything to me. Why would it have builder classes but not Table (i.e. an actual columnar container)?

**Comment by wesm** [comment]
@kszucs we aren't running the "large_memory" unit tests in Travis CI. What do you think about having a Docker target where we can run these so they can at least be spot-checked periodically?

File: cpp/src/arrow/array-binary-test.cc [hunk 2]
```diff
@@ -676,4 +678,104 @@ TEST_F(TestStringArray, TestSliceEquality) { CheckSliceEquality<BinaryType>(); }
...
+
+  const int iterations = 100;
+  for (int i = 0; i < iterations; ++i) {
+    random_bytes(bufsize, i, buf);
+    ASSERT_OK(builder_->Append(buf, bufsize));
+  }
+
+  ArrayVector chunks;
+  ASSERT_OK(builder_->Finish(&chunks));
+  ASSERT_EQ(iterations, static_cast<int>(chunks.size()));
```
Comment (pitrou) [inline cpp/src/arrow/array-binary-test.cc]: Can you also test the total data size or number of elements?

File: cpp/src/arrow/columnar/builder_binary.cc [hunk 3]
```diff
@@ -292,24 +276,50 @@ util::string_view FixedSizeBinaryBuilder::GetView(int64_t i) const {
...
+}
+
+Status ChunkedStringBuilder::Finish(ArrayVector* out) {
+  ArrayVector temp;
+  RETURN_NOT_OK(ChunkedBinaryBuilder::Finish(&temp));
 
+  out->clear();
+  out->reserve(temp.size());
+
+  // XXX: Change data type to string/utf8
```
Comment (pitrou) [inline cpp/src/arrow/columnar/builder_binary.cc]: Why is this an XXX?

File: cpp/src/arrow/compute/kernel.h [hunk 4]
```diff
@@ -61,19 +61,29 @@ struct ARROW_EXPORT Datum {
...
+      : value(value) {}
+  Datum(const std::shared_ptr<Table>& value)  // NOLINT implicit conversion
+      : value(value) {}
+  Datum(const std::vector<Datum>& value)  // NOLINT implicit conversion
+      : value(value) {}
+
+  // Cast from subtypes of Array to Datum
+  template <typename T,
+            typename = typename std::enable_if<std::is_base_of<Array, T>::value,
+                                               void>::type>
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernel.h]: I don't think `void` needs to be spelled out explicitly.

**Review by pitrou — COMMENTED** [review]
Partial review.

File: cpp/src/arrow/CMakeLists.txt (hunk 1 above)
Comment (wesm) [reply cpp/src/arrow/CMakeLists.txt]: Can you suggest another name? "core" perhaps?

Comment (pitrou) [reply cpp/src/arrow/CMakeLists.txt]: What do you intend to put in it?

Comment (wesm) [reply cpp/src/arrow/CMakeLists.txt]: I would like to have a place to organize the columnar data structure implementation details in a way that is logical for developers (of the Arrow codebase, not users of its public API) to navigate, and that promotes developer productivity (through easier navigation, smaller headers and compilation units, etc.). This should be unrelated to the development of compact public API (we should put a bit more effort into organizing that also -- it has grown a tad crufty, or at least it not well-documented)

The next thing to do would be to organize array.h/array.cc into similarly-named files. Parts of other files could be moved there, too. I couldn't bear to keep adding more classes to builder.h hence the decision to organize a bit here

Comment (pitrou) [reply cpp/src/arrow/CMakeLists.txt]: Call it `array` simply? We'd put all the `array-*` and `builder-*` stuff inside.

Comment (wesm) [reply cpp/src/arrow/CMakeLists.txt]: done

**Comment by wesm** [comment]
I'm all done here, just will make sure the build is passing

**Comment by wesm** [comment]
@xhochy @pitrou would you like to leave any more comments? The Python build timed out but otherwise this is ready to merge

File: cpp/src/arrow/array/builder_binary.cc [hunk 5]
```diff
@@ -79,23 +80,6 @@ Status BinaryBuilder::AppendNextOffset() {
   return offsets_builder_.Append(static_cast<int32_t>(num_bytes));
 }
 
-Status BinaryBuilder::Append(const uint8_t* value, int32_t length) {
```
Comment (pitrou) [inline cpp/src/arrow/array/builder_binary.cc]: I don't think it's useful to inline those if `AppendNextOffset` is not inlined.

File: cpp/src/arrow/pretty_print.cc [hunk 6]
```diff
@@ -19,7 +19,7 @@
 #include <cstdint>
 #include <iostream>
 #include <memory>
-#include <sstream>
+#include <sstream>  // IWYU pragma: keep
```
Comment (pitrou) [inline cpp/src/arrow/pretty_print.cc]: I keep seeing these in the code base. What are they for? (the "IWYU pragma" comments)

File: cpp/src/arrow/python/numpy_to_arrow.cc [hunk 7]
```diff
@@ -539,33 +539,31 @@ Status NumPyConverter::Visit(const BinaryType& type) {
 
   auto data = reinterpret_cast<const uint8_t*>(PyArray_DATA(arr_));
 
-  int item_length = 0;
+  auto AppendNotNull = [&builder, this](const uint8_t* data) {
+    // This is annoying. NumPy allows strings to have nul-terminators, so
+    // we must check for them here
+    int item_length = 0;
```
Comment (pitrou) [inline cpp/src/arrow/python/numpy_to_arrow.cc]: Could perhaps simply use the [`strnlen_s`](https://en.cppreference.com/w/c/string/byte/strlen) function.

File: cpp/src/arrow/pretty_print.cc (hunk 6 above)
Comment (wesm) [reply cpp/src/arrow/pretty_print.cc]: Have you run include-what-you-use before? The problem is that the C++ standard library is a little bit different on Windows, and so IWYU will say to remove a header sometimes and that will break the Windows build. So the pragmas are about having clean IWYU runs

Comment (pitrou) [reply cpp/src/arrow/pretty_print.cc]: I see, thanks. I think I ran it before, but I don't remember any specific issues.

Comment (wesm) [reply cpp/src/arrow/pretty_print.cc]: I think it's better to deal with this particular issue using mappings files (https://github.com/apache/arrow/tree/master/cpp/build-support/iwyu/mappings), so at some point we should try to clean up the "keep" pragmas using those

File: cpp/src/arrow/array/builder_binary.cc (hunk 5 above)
Comment (wesm) [reply cpp/src/arrow/array/builder_binary.cc]: Good point. I'll inline AppendNextOffset then

File: cpp/src/arrow/python/numpy_to_arrow.cc (hunk 7 above)
Comment (wesm) [reply cpp/src/arrow/python/numpy_to_arrow.cc]: is it ok to use `strnlen`? `strnlen_s` seems to be C11 only (though does C++11 imply that C11 extensions are available)?

Comment (pitrou) [reply cpp/src/arrow/python/numpy_to_arrow.cc]: Oh... I had read "C++11". Let's forget about it :-)

**Comment by wesm** [comment]
The inlining of BinaryBuilder methods produces a meaningful benchmark improvement

before

```
$ ./release/arrow-builder-benchmark --benchmark_filter=BinaryArray
2018-12-14 13:09:07
Running ./release/arrow-builder-benchmark
Run on (8 X 3700 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 256K (x4)
  L3 Unified 8192K (x1)
-------------------------------------------------------------------------------------
Benchmark                                              Time           CPU Iterations
-------------------------------------------------------------------------------------
BM_BuildBinaryArray/min_time:1.000                423367 us     423376 us          3   377.915MB/s
```

after

```
$ ./release/arrow-builder-benchmark --benchmark_filter=BinaryArray
2018-12-14 13:10:27
Running ./release/arrow-builder-benchmark
Run on (8 X 3700 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 256K (x4)
  L3 Unified 8192K (x1)
----------------------------------------------------------------------------------------
Benchmark                                                 Time           CPU Iterations
----------------------------------------------------------------------------------------
BM_BuildBinaryArray/min_time:1.000                   387697 us     387698 us          3   412.692MB/s
BM_BuildChunkedBinaryArray/min_time:1.000            370996 us     371003 us          4   431.263MB/s
```

**Comment by pitrou** [comment]
Nice :-)

**Comment by wesm** [comment]
+1

**Comment by kszucs** [comment]
@wesm created issue for running large memory tests [ARROW-4046](https://issues.apache.org/jira/browse/ARROW-4046)

**Comment by wesm** [comment]
thanks =)

**Comment by zygm0nt** [comment]
I still see this error when using 0.13.0, also tested with 0.12.0.

The code I've tested this with is the exact same code as in ARROW-4046:

```import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq


x = pa.array(list('1' * 2**30))

demo = 'demo.parquet'


def scenario():
    t = pa.Table.from_arrays([x], ['x'])
    writer = pq.ParquetWriter(demo, t.schema)
    for i in range(2):
        writer.write_table(t)
    writer.close()

    pf = pq.ParquetFile(demo)

    # pyarrow.lib.ArrowIOError: Arrow error: Invalid: BinaryArray cannot contain more than 2147483646 bytes, have 2147483647
    t2 = pf.read()

    # Works, but note, there are 32 row groups, not 2 as suggested by:
    # https://arrow.apache.org/docs/python/parquet.html#finer-grained-reading-and-writing
    tables = [pf.read_row_group(i) for i in range(pf.num_row_groups)]
    t3 = pa.concat_tables(tables)

scenario()```

**Comment by wesm** [comment]
You mean a different JIRA than https://issues.apache.org/jira/browse/ARROW-4046, right? Can you post this on the appropriate JIRA issue or create a new one so we can track this? Thanks

**Comment by zygm0nt** [comment]
Right, my mistake. I've meant this one: https://issues.apache.org/jira/browse/ARROW-3762

**Comment by yogeshg** [comment]
Hi @wesm , Big fan of your work!
Is there a corresponding issue in java? I could not find it. I am getting this error in spark for the same dataset that throws the above error in python.
```
SparkException: Job aborted due to stage failure: Task 9 in stage 5.0 failed 4 times, most recent failure: Lost task 9.3 in stage 5.0 (TID 255, 10.30.253.49, executor 4): org.apache.spark.sql.execution.QueryExecutionException: Encounter error while reading parquet files. One possible cause: Parquet column cannot be converted in the corresponding files. Details: 
    at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1$$anon$2.getNext(FileScanRDD.scala:259)
    at org.apache.spark.util.NextIterator.hasNext(NextIterator.scala:73)
    at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1$$anonfun$prepareNextFile$1.apply(FileScanRDD.scala:389)
    at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1$$anonfun$prepareNextFile$1.apply(FileScanRDD.scala:377)
    at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
    at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
    at org.apache.spark.util.threads.SparkThreadLocalCapturingRunnable$$anonfun$run$1.apply$mcV$sp(SparkThreadLocalForwardingThreadPoolExecutor.scala:93)
    at org.apache.spark.util.threads.SparkThreadLocalCapturingRunnable$$anonfun$run$1.apply(SparkThreadLocalForwardingThreadPoolExecutor.scala:93)
    at org.apache.spark.util.threads.SparkThreadLocalCapturingRunnable$$anonfun$run$1.apply(SparkThreadLocalForwardingThreadPoolExecutor.scala:93)
    at org.apache.spark.util.threads.SparkThreadLocalCapturingHelper$class.runWithCaptured(SparkThreadLocalForwardingThreadPoolExecutor.scala:63)
    at org.apache.spark.util.threads.SparkThreadLocalCapturingRunnable.runWithCaptured(SparkThreadLocalForwardingThreadPoolExecutor.scala:90)
    at org.apache.spark.util.threads.SparkThreadLocalCapturingRunnable.run(SparkThreadLocalForwardingThreadPoolExecutor.scala:93)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.parquet.io.ParquetDecodingException: Can not read value at 0 in block -1 in file XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
    at org.apache.parquet.hadoop.InternalParquetRecordReader.nextKeyValue(InternalParquetRecordReader.java:251)
    at org.apache.parquet.hadoop.ParquetRecordReader.nextKeyValue(ParquetRecordReader.java:207)
    at org.apache.spark.sql.execution.datasources.RecordReaderIterator.hasNext(RecordReaderIterator.scala:40)
    at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
    at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1$$anon$2.getNext(FileScanRDD.scala:228)
    ... 14 more
Caused by: java.lang.IllegalArgumentException: Illegal Capacity: -34
    at java.util.ArrayList.<init>(ArrayList.java:157)
    at org.apache.parquet.hadoop.ParquetFileReader$ConsecutiveChunkList.readAll(ParquetFileReader.java:1169)
    at org.apache.parquet.hadoop.ParquetFileReader.readNextRowGroup(ParquetFileReader.java:811)
    at org.apache.parquet.hadoop.InternalParquetRecordReader.checkRead(InternalParquetRecordReader.java:127)
    at org.apache.parquet.hadoop.InternalParquetRecordReader.nextKeyValue(InternalParquetRecordReader.java:222)
    ... 18 more

Driver stacktrace:
```

**Comment by wesm** [comment]
@yogeshg can you open a JIRA issue either in Arrow or Spark? I think that this is something that will have to be handled on the Spark side cc @BryanCutler

**Comment by yogeshg** [comment]
I'll do that soon.

**Comment by BryanCutler** [comment]
@yogeshg , this might be the related issue from the Java MR Parquet Reader that Spark uses https://issues.apache.org/jira/browse/PARQUET-980, but please open another JIRA if it is not

**Pull request closed without merging**